### PR TITLE
cmake: add llext compilation module

### PR DESCRIPTION
**Pull request by pillo79**
This PR abstracts the current llext compilation code used in the test to the `add_llext_target` function in the `cmake/modules/llext.cmake` module, integrating it with the current Zephyr build configuration as much as current llext code allows. 

Of course the "extension" model is designed to allow third party, independent binaries to be separately compiled and dynamically linked with the Zephyr binary at run time only. 
But having a fully integrated build of a Zephyr binary + a set of extensions has a number of advantages:

- Extensions can be built with the same compilation options of the main Zephyr binary (ABI, floating point, warnings etc), minimizing the risk of incompatibilities; 
- Extensions will be able to use all existing Zephyr macrobatics (such as devicetree or `zassert_*` macros) in their code. This will enable them to adapt to the current build requirements, and Twister tests will be able to influence test outcome from _within_ the extension code;
- @marc-hb suggested being able to build code with different licenses and distribution restrictions in the same single build, and distribute them as separate entities.

Some restrictions on the set of flags that can be used are automatically enforced. This list is expected to be reduced as the Zephyr llext linker improves.

-----
* 12cdb55e38e34ca404e821cda92f8f82f9c73be2 moves the current llext compilation code to an `add_llext_target` CMake function, so CMakeFiles for llext tests can be greatly simplified;
* 1be79094ecd041300fac9fe93578cabe2ca03b14 reworks `add_llext_target` to use a standard Zephyr library compile step for the llext object files, allowing to reuse include directories and macro definitions;
* bf2e24bf312e77e4f70543d982b024537b399624 completes the transition by moving all arch-specific flags to the cmake/compiler/ folder and reusing the toolchain compiler properties as much as possible.

### CODE REVIEW

File: subsys/llext/CMakeLists.txt [hunk 1]
```diff
@@ -2,4 +2,38 @@ if(CONFIG_LLEXT)
   zephyr_library()
   zephyr_library_sources(llext.c llext_export.c buf_loader.c)
   zephyr_library_sources_ifdef(CONFIG_LLEXT_SHELL shell.c)
+
+  # Compile a source .c file to an .llext file, using the appropriate
+  # commands and flags for the target architecture.
+  # Expects full paths to the source and target files.
+  # Additional flags can be provided via the LLEXT_C_FLAGS list.
+  function(llext_compile source target)
```
Comment (teburd) [inline subsys/llext/CMakeLists.txt:10]: Lets do this the right way, lets setup a target and add this as a cmake module under zephyr/cmake/modules/llext.cmake

It's absolutely needed and I appreciate the work here to add this.

Ultimately we may wish to provide some options using cmake variables as well for deciding compilation route (static vs got/plt linkage) as well as wanting to reuse as many other compilation flags as possible. All of that can live in the llext.cmake file then.

File: tests/subsys/llext/hello_world/CMakeLists.txt [hunk 2]
```diff
@@ -3,34 +3,20 @@
...
-endif()
+# Compile a simple hello world llext to an include file
+set(llext_src_file ${PROJECT_SOURCE_DIR}/src/llext/hello_world.c)
+set(llext_bin_file ${ZEPHYR_BINARY_DIR}/hello_world.llext)
+set(llext_inc_file ${ZEPHYR_BINARY_DIR}/include/generated/hello_world.inc)
 
-set(HELLO_WORLD_LLEXT ${PROJECT_BINARY_DIR}/hello_world.llext PARENT_SCOPE)
-
-add_custom_target(hello_world DEPENDS ${PROJECT_BINARY_DIR}/hello_world.llext)
+llext_compile(${llext_src_file} ${llext_bin_file})
```
Comment (teburd) [inline tests/subsys/llext/hello_world/CMakeLists.txt:21]: I'd recommend we rename this and make it closer to the builtin add_library/add_executable... like zephyr_add_llext(<name> [source1] [source2]...)

Which would create a target with the given <name> and an extension file <name>.llext as the output.

File: subsys/llext/CMakeLists.txt [hunk 3]
```diff
@@ -2,4 +2,38 @@ if(CONFIG_LLEXT)
...
+                -imacros ${AUTOCONF_H}
+                -o ${target}
+                ${source}
+      )
+    elseif(CONFIG_XTENSA)
+      list(PREPEND LLEXT_C_FLAGS "-shared" "-fPIC" "-nostdlib" "-nodefaultlibs")
+      get_filename_component(target_dir ${target} DIRECTORY)
+      get_filename_component(target_name ${target} NAME_WE)
+      set(pre_target ${target_dir}/${target_name}.pre.llext)
+      add_custom_command(OUTPUT ${target}
```
Comment (teburd) [inline subsys/llext/CMakeLists.txt:25]: this lacks the add_custom_target bit

**Review by teburd — COMMENTED** [review]
Some minor nits about where this lives, how its named, and what it does. I love the path this is going.

See https://github.com/thesofproject/sof/pull/8180 and the comments by @marc-hb for additional commentary on why maybe some of this is needed.

**Comment by pillo79** [comment]
Thanks for the pointers @teburd, will look into it tomorrow!

File: subsys/llext/CMakeLists.txt (hunk 1 above)
Comment (teburd) [reply subsys/llext/CMakeLists.txt:10]: @tejlmand any thoughts on this?

**Comment by pillo79** [comment]
@teburd I have addressed all notes, but I was not able to create the perfect `zephyr_add_llext` clone as I couldn't find out how to make the CMake generators (esp. `$<TARGET_OBJECTS:target>`) work with custom targets, so figuring out the full path to the output llext was clumsy.

Ended up defining an `add_llext_target` that mostly matches `add_custom_target`, and can be used like this:
```C
add_llext_target(hello_world_llext
  OUTPUT  ${PROJECT_BINARY_DIR}/hello_world.llext
  SOURCES ${PROJECT_SOURCE_DIR}/src/llext/hello_world.c
  C_FLAGS -Werror
)
```
so precise path control is still in the hands of the developer (I have skimmed the SOF PR and I see that would be helpful).

This now uses the default Zephyr compile step for the actual llext sources, which should allow a lot more flexibility in llext code. It was a pain to make the Xtensa port work though, since I found no way to create a dynamic library (the Zephyr toolchain forbids an `add_library(... SHARED ...)`) - I had to compile as an OBJECT library and then manually invoke the compiler to create a shared object. There is no change in the test output, but @lyakh, please try if you have issues updating SOF to this and let me know.

**Comment by pillo79** [comment]
> Lets do this the right way, lets setup a target and add this as a cmake module under zephyr/cmake/modules/llext.cmake

Missed this bit! Will move the `add_lext_target` code there ASAP.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for improving on this.

Some comments / observations from a first review round, but on a more general level I would like to know better the general intentions, perhaps some more documentation.

The llext feature was introduced here: https://github.com/zephyrproject-rtos/zephyr/pull/57086

As I see this commit it tries to improve the llext work 👍 by:
- adding more tests / better test coverage
- creating Zephyr specific functions for creation of llext files.

Unfortunately it's not fully clear to me which parts relates to improved testing, and which parts is for improved llext functionality in itself.

As I understand `llext`, then ll code is generally expected to be built independently of the actual Zephyr build (because if it was built as part of the Zephyr build there would be little reason to use the llext feature).
But the `add_llext_target()` is created as part of a Zephyr build.

Also I notice that the `add_llext_target()` tries to use the `zephyr_interface`, for example as:  `$<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_DEFINITIONS>`.
My question therefore is: is a ll module expected to be built with same flag as Zephyr itself ?
I would assume some flags are important, like fp support, but others doesn't matter.
But if it is expected that users can build a ll module using Zephyr infrastructure (and re-use flags from a given Zephyr), but without building  Zephyr itself, then I think a different principle is needed.

Also I think an update to the llext documentation would be nice.

File: tests/subsys/llext/Kconfig [hunk 4]
```diff
@@ -7,6 +7,6 @@ source "Kconfig.zephyr"
 
 config LLEXT_TEST_HELLO
 	tristate "llext hello test"
-	default n
+	default y
 	help
-	  Set to "m" to test hello-world loading
+	  Set to "m" to test hello-world loading when modules are enabled.
```
[the remark this answers is not in the document]
Comment (pillo79) [reply tests/subsys/llext/Kconfig:12]: The original code for that file was from #67105. Until that got merged, Zephyr did not support `tristate` entries, so a quick example was added to that llext test to showcase the new functionality. Basically this is the only place where you can set a config value to `m` and expect it to work. The CMakeFile actually tests this functionality by requiring that if tristates (aka `CONFIG_MODULES`) are enabled, this is set to "m".

Would something like this be OK?

> This enables building the hello_world test case. Note that if CONFIG_MODULES is enabled you should set this to "m".

File: tests/subsys/llext/testcase.yaml [hunk 5]
```diff
@@ -1,22 +1,25 @@
...
-  llext.simple.xtensa:
-    arch_allow: xtensa
+      - arch:arm:CONFIG_ARM_MPU=n
+      - CONFIG_LLEXT_STORAGE_WRITABLE=n
+  llext.simple.writable:
     extra_configs:
+      - arch:arm:CONFIG_ARM_MPU=n
       - CONFIG_LLEXT_STORAGE_WRITABLE=y
+  llext.simple.modules_enabled:
+    arch_allow: xtensa # only useful to test one arch, Xtensa is faster
```
[the remark this answers is not in the document]
Comment (pillo79) [reply tests/subsys/llext/testcase.yaml]: (that is definitely only a time saver to cut the number of tests down, and having to choose, I picked the one compiling faster :slightly_smiling_face:)

File: subsys/llext/CMakeLists.txt [hunk 6]
```diff
@@ -1,5 +1,84 @@
...
+    add_custom_target(${target_name} ALL
+      COMMENT "Compiling ${output_name}"
+      DEPENDS ${output_file}
+    )
+
+    # Compile the source file to an .llext file
+    if(CONFIG_ARM)
+      list(PREPEND LLEXT_C_FLAGS "-mlong-calls" "-mthumb")
+      add_custom_command(OUTPUT ${output_file}
+        COMMAND ${CMAKE_C_COMPILER} ${LLEXT_C_FLAGS} -c
```
[the remark this answers is not in the document]
Comment (pillo79) [reply subsys/llext/CMakeLists.txt]: :+1: I will add gcc properties to the target_specific files in `cmake/compiler/gcc`.

File: subsys/llext/CMakeLists.txt [hunk 7]
```diff
@@ -1,5 +1,84 @@
...
+    if(NOT source_count EQUAL 1)
+      message(FATAL_ERROR "add_llext_target: only one source file is supported")
+    endif()
+
+    set(output_file ${LLEXT_OUTPUT})
+    set(source_file ${LLEXT_SOURCES})
+    get_filename_component(output_name ${output_file} NAME)
+
+    # Add user-visible target and dependency
+    add_custom_target(${target_name} ALL
```
[the remark this answers is not in the document]
Comment (pillo79) [reply subsys/llext/CMakeLists.txt]: ALL is probably redundant here. I will remove it so it only gets made when needed to create the source for the test code.

I agree building Zephyr and extensions them separately is a very important use case. However, there are 2 instances off the top of my head where it makes sense to have them integrated into a single build run:
* test cases, such as the one here :slightly_smiling_face: 
* (going out on a limb, maybe @lyakh can confirm) e.g. for SOF to build and release a new audio firmware + a set of filters.

File: subsys/llext/CMakeLists.txt [hunk 8]
```diff
@@ -1,5 +1,84 @@
...
+  # Example usage:
+  #   add_llext_target(hello_world
+  #     OUTPUT  ${PROJECT_BINARY_DIR}/hello_world.llext
+  #     SOURCES ${PROJECT_SOURCE_DIR}/src/llext/hello_world.c
+  #     C_FLAGS -Werror
+  #   )
+  # will compile the source file src/llext/hello_world.c to a file
+  # ${PROJECT_BINARY_DIR}/hello_world.llext, adding -Werror to the compilation.
+  #
+  function(add_llext_target target_name)
```
[the remark this answers is not in the document]
Comment (pillo79) [reply subsys/llext/CMakeLists.txt]: That has now moved to `cmake/modules/llext.cmake`. Is this new location acceptable?

**Comment by pillo79** [comment]
Thanks @tejlmand for the in-depth review and insights. I will address them and provide a revised version. 

Regarding your general questions on llext, there is still a lot of work to do before this can be made easy to use  (currently, global and static variables are mostly unhandled... :sweat_smile:). I already have some of these linker features working in a draft PR, but being able to test them was equally important, therefore I started on this "easier" road.

This series does not add any test code or llext feature, it just refactors `tests/subsys/llext` so an llext test is cleaner and simpler to replicate (removing per-arch YAML tests, providing a generic build machinery, etc).

> is a ll module expected to be built with same flag as Zephyr itself ?

It may be or may be not. Of course a fully decoupled, independent build is possible, but having a sort of "overlay" with tightly integrated Zephyr code may be really useful. For example, my next PR will include a test llext that uses assertions _from the llext_. This requires most of the flags, include directories etc to be lifted from the main build directly to the extension file.

I promise I will address the appalling documentation as soon as some more features are in. :crossed_fingers:

**Comment by teburd** [comment]
> Thanks @tejlmand for the in-depth review and insights. I will address them and provide a revised version.
> 
> Regarding your general questions on llext, there is still a lot of work to do before this can be made easy to use (currently, global and static variables are mostly unhandled... 😅). I already have some of these linker features working in a draft PR, but being able to test them was equally important, therefore I started on this "easier" road.
> 
> This series does not add any test code or llext feature, it just refactors `tests/subsys/llext` so an llext test is cleaner and simpler to replicate (removing per-arch YAML tests, providing a generic build machinery, etc).
> 
> > is a ll module expected to be built with same flag as Zephyr itself ?

I would like to see this happen yes, most of the compilation flags should be replicated with perhaps a few differences. Mostly because of the way llext's need to be built today as partially linked elfs for arm with limited relocations supported. It would be great to replicate most of the warning flags, optimization level, call abi, various math flags, etc. I'm glad to see this is being worked out as I puzzled over how to to reuse the zephyr compiler flags myself, they aren't a simple cmake variables as I gather but instead attached to a target (zephyr_interface) and this frankly made things a little awkward with my limited cmake hackery.

> 
> It may be or may be not. Of course a fully decoupled, independent build is possible, but having a sort of "overlay" with tightly integrated Zephyr code may be really useful. For example, my next PR will include a new test llext that uses `zassert_equal` _from inside the llext_. This requires most of the flags, include directories etc to be lifted from the main build directly to the extension file.

I have one user that will most likely need to build things outside of zephyr. Related is the compiler flags which I'd like to export into a generated set of files, @edersondisouza might have better ideas here though perhaps. But imagine the scenario of base firmware being created alongside a plugin/extension/whatever SDK including headers, cmake files, linker scripts, etc etc.

> 
> I promise I will address the appalling documentation as soon as some more features are in. 🤞

Documentation improvements are very much appreciated.

File: tests/subsys/llext/testcase.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pillo79) [reply tests/subsys/llext/testcase.yaml]: Awesome feature, thanks for pointing this out! :bow: 
Will use `platform_key: simulation` instead. :+1:

File: cmake/modules/llext.cmake [hunk 9]
```diff
@@ -0,0 +1,111 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tejlmand) [inline cmake/modules/llext.cmake:1]: Not related to this specific line, but comment relating to https://github.com/zephyrproject-rtos/zephyr/pull/67431/commits/ee9a514531497917b596a65a97d14865d0e1acf9 + https://github.com/zephyrproject-rtos/zephyr/pull/67431/commits/113902e85e9dbb0e35e658906fdefb8a4a4deed3 vs https://github.com/zephyrproject-rtos/zephyr/pull/67431/commits/d435f7b0c0e31d83e445db3dc5e5a09250240f8f: 
reviewing PR's where code is moved around between commits makes it very hard to review a PR commit-by-commit.

File: subsys/llext/CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (tejlmand) [reply subsys/llext/CMakeLists.txt]: yes, I noticed later, and actually made a comment, but seems I missed adding the comment to the review 😞 .
Please see followup comment here : https://github.com/zephyrproject-rtos/zephyr/pull/67431#discussion_r1450947598

**Comment by pillo79** [comment]
Thanks everyone for the feedback, I should have addressed all concerns. The commits have also been reworked to avoid file moves as much as possible.

File: cmake/modules/llext.cmake [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      COMMAND $<TARGET_PROPERTY:bintools,strip_command>
+              $<TARGET_PROPERTY:bintools,strip_flag>
+              $<TARGET_PROPERTY:bintools,strip_flag_remove_section>.xt.*
+              $<TARGET_PROPERTY:bintools,strip_flag_infile>${pre_output_file}
+              $<TARGET_PROPERTY:bintools,strip_flag_outfile>${output_file}
+              $<TARGET_PROPERTY:bintools,strip_flag_final>
+      DEPENDS ${target_name}_lib $<TARGET_OBJECTS:${target_name}_lib>
+    )
+
+  endif()
```
Comment (marc-hb) [inline cmake/modules/llext.cmake:129]:
```suggestion
  else()
  message(FATAL_ERROR "unsupported...
  endif()
```

File: tests/subsys/llext/hello_world/CMakeLists.txt [hunk 11]
```diff
@@ -3,38 +3,32 @@
...
-		${PROJECT_SOURCE_DIR}/hello_world.c
-	COMMAND ${CROSS_COMPILE}strip -R .xt.*
-		-o ${PROJECT_BINARY_DIR}/hello_world.llext
-		${PROJECT_BINARY_DIR}/hello_world.pre.llext
-    )
-endif()
+project(llext_hello_world_test)
+
+if(NOT CONFIG_LLEXT_TEST_HELLO)
+  # test disabled, nothing to do
```
Comment (marc-hb) [inline tests/subsys/llext/hello_world/CMakeLists.txt:9]: Pardon my ignorance but is this really how/where tests get enabled/disabled? In Kconfig and in their CMakeLists.txt file? This looks strange, I would have expected `testcase.yaml` to be enough...

**Review by marc-hb — COMMENTED** [review]
First, I want to second @teburd and state how grateful we are that you are doing this. It's incredibly rare to find people with the CMake skills and patience to do something like this (not me!) which makes this PR even more valuable. This is amazing.

Answering more than one single person below:


> Unfortunately it's not fully clear to me which parts relates to improved testing, and which parts is for improved llext functionality in itself.

Also, tests are only a means to an end. The most valuable part of this PR is IMHO NOT the tests but the generic CMake code. Please update the currently unassuming PR title to reflect that! Drop the word "tests" from it. Ideally, every code addition or change should have corresponding tests; this should go without even saying :-)


> I agree building Zephyr and extensions them separately is a very important use case. However, there are 2 instances off the top of my head  where it makes sense to have them integrated into a single build run...

Maybe add this one:  you can have a mix of sources _with different licences_ in a single build. These different licences may require you to release binaries from the same build separately. Building everything at once make sure everything is in sync and matches: sources, build flags, toolchains, etc.

Of course building separately is desirable too - but not just.

> > Due to some file moves it is best to review each commit independently:

> reviewing PR's where code is moved around between commits makes it very hard to review a PR commit-by-commit.

Have you considered splitting this into smaller PRs? They could be merged faster.

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

> I promise I will address the appalling documentation as soon as some more features are in.

Hey, maybe something I can help with in the future? I would recommend not spending much time on it yet, better wait for the dust to settle a bit first. `llext` seems very young yet, probably does not need a huge crowd of users yet...

File: tests/subsys/llext/hello_world/CMakeLists.txt (hunk 11 above)
Comment (pillo79) [reply tests/subsys/llext/hello_world/CMakeLists.txt:9]: That code was introduced in aee2d1a677 as the only example usage of `tristate` in the Zephyr code. There is no actual need to do that, but I did not want to remove it either, so I just moved that around.

**Review by teburd — DISMISSED** [review]

**Review by marc-hb — DISMISSED** [review]
My CMake expertise is just above the average (which is itself close to zero) but this looks very clean, readable and maintainable. thanks! Great starting point for sure.

If @tejlmand, @teburd and @lyakh are happy then I am too.

**Comment by tejlmand** [comment]
> I have one user that will most likely need to build things outside of zephyr. Related is the compiler flags which I'd like to export into a generated set of files, @edersondisouza might have better ideas here though perhaps.

Zephyr already supports https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MAKEFILE_EXPORTS.
That feature exports Zephyr's build flags to files named: `Makefile.exports.<lang>`

That feature goes a long time back, #22020, and maybe we should consider to support an equivalent `CONFIG_CMAKE_BUILD_FLAGS_EXPORTS` setting which does the same, but in a format which can be directly consumed by another CMake build system.
Perhaps with some more configuration info.

As another piece of info I can mention that Zephyr default enables `CMAKE_EXPORT_COMPILE_COMMANDS`, although that feature probably contains a bit more info than what is needed in this case.

Feel free to open an enhancement issue describing the exact needs / use-case, and feel free to refer to some of the existing features I mention above.

I even think providing such a feature could allow an even have cleaner split between the build of Zephyr and the build of llext module in the future.

**Review by tejlmand — CHANGES_REQUESTED** [review]
implementation looks nice.

A single comment with regards to the new `llext.cmake` module, which I would like to address.
Both to get the reasoning described, but also for future reference as the reason for allowing an exception in the llext case.

File: cmake/modules/llext.cmake [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+# Example usage:
+#   add_llext_target(hello_world
+#     OUTPUT  ${PROJECT_BINARY_DIR}/hello_world.llext
+#     SOURCES ${PROJECT_SOURCE_DIR}/src/llext/hello_world.c
+#     C_FLAGS -Werror
+#   )
+# will compile the source file src/llext/hello_world.c to a file
+# ${PROJECT_BINARY_DIR}/hello_world.llext, adding -Werror to the compilation.
+#
+function(add_llext_target target_name)
```
[the remark this answers is not in the document]
Comment (pillo79) [reply cmake/modules/llext.cmake:29]: Right! I was not sure about merging this into `extensions.cmake` for two reasons:
- llext is quite young and I assume a lot of work will be done on that function (and possibly others), so I wanted to make it clear that work on that file was not interfering with the rest of Zephyr developments;
- the functions in `extensions.cmake` are absolutely generic and really useful to every part of Zephyr, 
while llext is (currently!) a very specific functionality, so it looked early to have this so front and center.

Still, no fundamental objections to moving, of course. If you wish so, should I create a "7. Linkable loadable extensions" section at the end?

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/modules/llext.cmake:29]: thanks for the reasoning.

While it's true that majority of functions in `extensions.cmake` are generic / useful to most parts of Zephyr development, then there does exists some functions which are more dedicated to certain areas.

Although young, then let's consider the llext feature an integral part of the Zephyr build system, and giving it is own section in `extensions.cmake` 🎉

**Review by tejlmand — APPROVED** [review]

**Review by teburd — APPROVED** [review]

File: cmake/modules/extensions.cmake [hunk 13]
```diff
@@ -4786,3 +4787,138 @@ macro(zephyr_check_flags_exclusive function prefix)
...
+  set(zephyr_flags
+      "$<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_OPTIONS>"
+  )
+  set(zephyr_filtered_flags
+      "$<FILTER:${zephyr_flags},EXCLUDE,${llext_remove_flags_regexp}>"
+  )
+
+  # Compile the source file to an object file using current Zephyr settings
+  # but a different set of flags
+  add_library(${target_name}_lib OBJECT ${source_file})
```
Comment (teburd) [inline cmake/modules/extensions.cmake:4868]: I did try this before but it didn't quite give me what I wanted, every .c turns into a .o and the subsequent custom command would fail. Did you try this with multiple .c files representing an extension?

The llext would be constructed using a linker to merge the elfs in the case of arm then.

Comment (pillo79) [reply cmake/modules/extensions.cmake:4868]: There is a check that forces only a single file at this moment exactly for that reason - we don't currently link the files in the ARM case, so we would have to deal with a bunch of unrelated objects. To be able to load several files, we need to partially link them at build time, then enable dynamic linking for ARM. That is the goal of #66553 (still very much WIP, needs rebase and splitting in smaller PRs).
We could also link between objects in llext at runtime, but I'm not in favor of this. I think we should only link with a few exported symbols and not handle every intra-module symbol reference.

File: cmake/modules/extensions.cmake [hunk 14]
```diff
@@ -4845,34 +4845,66 @@ function(add_llext_target target_name)
...
-              ${source_file}
+              $<TARGET_OBJECTS:${target_name}_lib>
       COMMAND ${CROSS_COMPILE}strip -R .xt.*
               -o ${output_file}
               ${pre_output_file}
-      DEPENDS ${source_file}
+      DEPENDS ${target_name}_lib $<TARGET_OBJECTS:${target_name}_lib>
     )
-  else()
-    message(FATAL_ERROR "add_llext_target: unsupported architecture")
```
Comment (marc-hb) [inline cmake/modules/extensions.cmake]: Did you really mean to drop this line?

File: cmake/modules/extensions.cmake [hunk 15]
```diff
@@ -4801,9 +4801,13 @@ endmacro()
...
-# Arch-specific flags will be added automatically. The C_FLAGS argument
-# can be used to pass additional compiler flags to the compilation of
-# the source file.
+# The llext code will be compiled with mostly the same C compiler flags used
+# in the Zephyr build, but with some important modifications. The list of
+# flags to remove and flags to append is controlled respectively by the
+# LLEXT_REMOVE_FLAGS and LLEXT_APPEND_FLAGS global variables.
+
+# The C_FLAGS argument can be used to pass additional compiler flags to the
+# compilation of this particular llext.
```
Comment (marc-hb) [inline cmake/modules/extensions.cmake:4810]: in CI we use _command-line_ parameters like `-DEXTRA_CXXFLAGS=Werror` a lot. Will they work with `llext` too?

File: cmake/modules/extensions.cmake [hunk 16]
```diff
@@ -4845,14 +4849,19 @@ function(add_llext_target target_name)
...
+  list(TRANSFORM LLEXT_REMOVE_FLAGS
+       REPLACE "(.+)" "^\\1$"
+       OUTPUT_VARIABLE llext_remove_flags_regexp
+  )
+  string(REPLACE ";" "|" llext_remove_flags_regexp "${llext_remove_flags_regexp}")
+  set(zephyr_flags
+      "$<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_OPTIONS>"
+  )
+  set(zephyr_filtered_flags
+      "$<FILTER:${zephyr_flags},EXCLUDE,${llext_remove_flags_regexp}>"
```
Comment (marc-hb) [inline cmake/modules/extensions.cmake:4863]: This is scary stuff :-) How about adding a couple `message(DEBUG $llext_remove_flags_regexp)` and `message(DEBUG zephyr_filtered_flags)`?
I'm surprised you didn't need them for yourself!

File: cmake/modules/extensions.cmake [hunk 17]
```diff
@@ -4897,14 +4906,19 @@ function(add_llext_target target_name)
...
+              $<TARGET_PROPERTY:bintools,strip_flag>
+              $<TARGET_PROPERTY:bintools,strip_flag_remove_section>.xt.*
+              $<TARGET_PROPERTY:bintools,strip_flag_infile>${pre_output_file}
+              $<TARGET_PROPERTY:bintools,strip_flag_outfile>${output_file}
+              $<TARGET_PROPERTY:bintools,strip_flag_final>
       DEPENDS ${target_name}_lib $<TARGET_OBJECTS:${target_name}_lib>
     )
 
+  else()
+    message(FATAL_ERROR "add_llext_target: unsupported architecture")
```
Comment (marc-hb) [inline cmake/modules/extensions.cmake:4922]: Removed and added back, looks like a small git glitch.

**Review by marc-hb — APPROVED** [review]

File: cmake/modules/extensions.cmake (hunk 15 above)
Comment (pillo79) [reply cmake/modules/extensions.cmake:4810]: I am no CMake guru so I just tested - it does (well, with EXTRA_CFLAGS as it's a C file, but it does indeed use command line flags in the llext as well).

File: cmake/modules/extensions.cmake (hunk 16 above)
Comment (pillo79) [reply cmake/modules/extensions.cmake:4863]: I used plenty of them for debugging indeed :innocent: but I did not push them, as they are quite cumbersome: to see the end result to that gibberish you need to add _a `custom_target` that echoes the flags on the console_. The generator syntax is next-level painful to me!

Comment (marc-hb) [reply cmake/modules/extensions.cmake:4863]: Ah yes of course, it's build-time. Got it.

Comment (teburd) [reply cmake/modules/extensions.cmake:4863]: cmake is next level painful, thanks for taking this on

File: cmake/modules/extensions.cmake [hunk 18]
```diff
@@ -4786,3 +4787,138 @@ macro(zephyr_check_flags_exclusive function prefix)
...
+#
+function(add_llext_target target_name)
+  set(single_args OUTPUT)
+  set(multi_args SOURCES;C_FLAGS)
+  cmake_parse_arguments(PARSE_ARGV 1 LLEXT "${options}" "${single_args}" "${multi_args}")
+
+  # Check that the llext subsystem is enabled for this build
+  if (NOT CONFIG_LLEXT)
+    message(FATAL_ERROR "add_llext_target: CONFIG_LLEXT must be enabled")
+  endif()
```
Comment (kartben) [inline cmake/modules/extensions.cmake:4829]: Just curious, what would be wrong with trying/wanting to generate an llext for a project that doesn't have LLEXT enabled. One needn't *use* LLEXT in a project to be able to have "snippets" of it built as llext, no?

Comment (pillo79) [reply cmake/modules/extensions.cmake:4829]: Well, indeed you could compile only the llext, since there's no real technical barrier. However, the generated file would not work with the same Zephyr core with which it was built (no exported symbols or machinery to load it), so it looked like a config error in my eyes.

---

Now that I look at it again, the converse is useful though - if you know what you are doing and _purposefully disable_ `CONFIG_LLEXT` on a project that defines such targets, it makes sense not to build them _and not complain_, unless they are required by other parts in the project.

Taking into account your other comment on dependencies ("why isn't a llext built if it's defined as part of a project?"), I think the most appropriate way to deal with all this should be:
* if `CONFIG_LLEXT` is not enabled, do not complain in `add_llext_target` and define an empty target instead, so other `llext_`* APIs don't break as well.
* automatically add the llext target to ALL so it's always built if it's defined in the CMake.

This would result in an actual build error only when LLEXT is disabled and the output file is a required dependency (for example by `generate_inc_file_for_target`) - or if the Zephyr code actually calls llext APIs, obviously.

@teburd, what do you think about this?

Comment (marc-hb) [reply cmake/modules/extensions.cmake:4829]: Sorry for the belated reply and - much worse - sorry for not being @teburd.

> disable CONFIG_LLEXT on a project that defines such targets, it makes sense not to build them and not complain,

I probably (!) disagree with that? The main purpose of Kconfig is to turn software features on and off. But "llext" isn't really a software _feature_, it's more about HOW you want some of those features to be delivered: as a separate module or built-in? So CONFIG_LLEXT is more like a "dependency" of real features.

Consider the Linux kernel, which many `llext` aspects take inspiration from. It had to already deal with this problem. If you turn off Linux CONFIG_MODULES in `make menuconfig` then it immediately forces you to make all features built-in. Editing the .config and running `make oldefconfig` does the same thing.

Of course there's a massive difference: the Linux kernel can very quickly easily flip drivers that support between module and built-in, whereas `llext` does not do that all. But "inspiration" is still useful IMHO.

So if you:
- enable some particular llext feature in Kconfig, AND
- you turn off CONFIG_LLEXT

... then you have a **configuration error** and the build should 1) fail, 2) fail as soon as possible, not wait until compilation whether some symbol is used maybe, maybe not.

In other words, if you turn off CONFIG_LLEXT then you MUST also turn off all features that depend on it - and maybe Kconfig can/should help with that but either way the build should fail ASAP if you give it an inconsistent configuration.

Comment (marc-hb) [reply cmake/modules/extensions.cmake:4829]: > I'll try to draft one shortly and share it so we can all agree on the long term path

Thank you so much for leading here again. Whatever are the use cases and intended Kconfig interactions, I think we all agree they cannot just be inferred from scattered comments in source and code reviews. Some high-level design documentation is required indeed.
